include/settings.hpp

```diff
@@ -0,0 +1,428 @@
+            for (auto &[key, value] : invalid_settings) {
+                fmt::print(stderr, "key: {} - value-type: {}\n", key, value.index());
+            }
+        }
+    }
+
+    [[nodiscard]] tag_t::map_type
+    set(const tag_t::map_type &parameters, SettingsCtx = {}) {
+        tag_t::map_type ret;
+        if constexpr (refl::is_reflectable<Node>()) {
+            std::lock_guard lg(_lock);
+            for (const auto &[localKey, localValue] : parameters) {
+                const auto &key    = localKey;
+                const auto &value  = localValue;
+                bool        is_set = false;
+                for_each(refl::reflect(*settings_base<Node>::_node).members, [&, this](auto member) {
+                    using Type = std::remove_cvref_t<decltype(member(*settings_base<Node>::_node))>;
+                    if constexpr (is_writable(member) && (std::is_arithmetic_v<Type> || std::is_same_v<Type, std::string>) ) {
+                        if (std::string(get_display_name(member)) == key && std::holds_alternative<Type>(value)) {
+                            if (_auto_update.contains(key)) {
+                                _auto_update.erase(key);
+                            }
+                            _staged.insert_or_assign(key, value);
+                            settings_base<Node>::_changed.store(true);
+                            is_set = true;
+                        }
+                    }
+                });
+                if (!is_set) {
+                    ret.insert_or_assign(key, pmtv::pmt(value));
+                }
+            }
+        }
+
+        return ret; // N.B. returns those <key:value> parameters that could not be set
+    }
+
+    void
+    auto_update(const tag_t::map_type &parameters, SettingsCtx = {}) {
+        if constexpr (refl::is_reflectable<Node>()) {
+            for (const auto &[localKey, localValue] : parameters) {
+                const auto &key   = localKey;
+                const auto &value = localValue;
+                for_each(refl::reflect(*settings_base<Node>::_node).members, [&](auto member) {
+                    using Type = std::remove_cvref_t<decltype(member(*settings_base<Node>::_node))>;
+                    if constexpr (is_writable(member) && (std::is_arithmetic_v<Type> || std::is_same_v<Type, std::string>) ) {
+                        if (std::string(get_display_name(member)) == key && std::holds_alternative<Type>(value)) {
+                            _staged.insert_or_assign(key, value);
+                            settings_base<Node>::_changed.store(true);
+                        }
+                    }
+                });
+            }
+        }
+    }
+
+    [[nodiscard]] const tag_t::map_type &
+    staged_parameters() const noexcept {
+        std::lock_guard lg(_lock);
+        return _staged;
+    }
```

Comment:
By returning the reference to `_staged` the access to it outlives the scope of the lock guard. It's a const ref, so the current thread cannot change the it, but another thread could call `set(...)` while the current thread is reading from the map.